native/src/main/java/io/ballerina/stdlib/xmldata/XmlToJson.java

```diff
@@ -202,13 +203,15 @@ private static void processAttributeWithAnnotation(BXmlItem xmlItem, String attr
                 } else if (fieldType instanceof UnionType) {
                     UnionType bUnionType = (UnionType) fieldType;
                     for (Type memberType : bUnionType.getMemberTypes()) {
-                        if (memberType.getTag() == TypeTags.RECORD_TYPE_TAG)  {
+                        if (memberType.getTag() == TypeTags.TYPE_REFERENCED_TYPE_TAG)  {
                             annotationType = memberType;
                         }
                     }
                 }
                 if (annotationType.getTag() == TypeTags.RECORD_TYPE_TAG) {
                     annotations = ((RecordType) annotationType).getAnnotations();
+                } else if (annotationType.getTag() == TypeTags.TYPE_REFERENCED_TYPE_TAG) {
+                    annotations = ((RecordType) ((ReferenceType) annotationType).getReferredType()).getAnnotations();
```

Comment:
What if this is a recursive reference type?

Reply:
Updated to support both scenarios in [7f64a6e](https://github.com/ballerina-platform/module-ballerina-xmldata/pull/465/commits/7f64a6e596abb50fe0e020535f83f51d06d4941d)